### PR TITLE
AWQ minor performance improvements to smoothing

### PR DESCRIPTION
**Pull request by brian-dellabetta**
SUMMARY:
I wanted to investigate how the runtime for `apply_smoothing` in AWQ could be decreased with torch.compile and removing python for loops. These changes reduce the runtime of `apply_smoothing` by about 8-10%, by vectorizing the loss computation.

TEST PLAN:
Confirmed wikitext PPL scores are virtually the same when running AWQ on `"meta-llama/Llama-3.2-3B-Instruct"` and `"Qwen/Qwen2.5-7B-Instruct"`

### CODE REVIEW

File: src/llmcompressor/modifiers/awq/base.py [hunk 1]
```diff
@@ -541,25 +541,35 @@ def smooth(module):
             v.batch_intermediates.clear()
         self._assert_all_activations_consumed()
 
-    def _run_samples(self, module: Module) -> List[torch.Tensor]:
+    def _get_flattened_output(self, module: Module) -> torch.Tensor:
+        """
+        Returns output of running cached batch inputs through module
+        Output tensor is 1D, as shapes aren't necessary for calculating loss
+        """
```
[the remark this answers is not in the document]
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py]: added

File: src/llmcompressor/modifiers/awq/base.py [hunk 2]
```diff
@@ -677,6 +659,17 @@ def _assert_all_activations_consumed(self):
             raise RuntimeError("Some cached activations were not used")
 
 
+@torch.no_grad()
+@torch.compile()
+def _compute_loss(
+    fp16_output: torch.Tensor,
+    int_w_output: torch.Tensor,
+) -> torch.Tensor:
+    """Compute MSE loss for each batch"""
```
[the remark this answers is not in the document]
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py]: added

File: src/llmcompressor/modifiers/awq/base.py [hunk 3]
```diff
@@ -677,6 +660,19 @@ def _assert_all_activations_consumed(self):
...
+@torch.no_grad()
+@torch.compile()
+def _compute_loss(
+    fp16_output: torch.Tensor,
+    int_w_output: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Compute MSE loss over the flattened output of all batches
+    """
+    return (fp16_output - int_w_output).view(-1).float().pow(2).mean()
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py]: Using `torch.nn.MSE()` here might be faster and likely already uses a precompiled kernel

File: src/llmcompressor/modifiers/awq/base.py [hunk 4]
```diff
@@ -677,6 +660,19 @@ def _assert_all_activations_consumed(self):
...
+
+
+@torch.compile()
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:668]: I wonder if torch.compile performs better without the asserts?

File: src/llmcompressor/modifiers/awq/base.py (hunk 3 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py]: good call, it's pretty much the same runtime-wise but a lot cleaner code-wise, we don't need a custom function. updated

File: src/llmcompressor/modifiers/awq/base.py (hunk 4 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:668]: I tried just now with asserts commented out, seems to shave off a few seconds. i think the if statements might not make it not all that beneficial? I think i'll leave the asserts as is, will be helpful for debugging. we can always add in future

Comment (kylesayrs) [reply src/llmcompressor/modifiers/awq/base.py:668]: Tracing has asserts turned off by default
https://github.com/pytorch/pytorch/blob/main/torch/fx/proxy.py#L131

I wasn't sure about torch.compile

**Review by kylesayrs — DISMISSED** [review]

**Comment by brian-dellabetta** [comment]
Closing this as it is pretty stale, and as of v0.9.0 the batched calibration runs provide a significant speedup to AWQ far surpassing minor improvements of this branch
